Step2 - 문자열 덧셈 계산기

stevejkang opened the pull request:
안녕하세요! Step2 - 문자열 덧셈 계산기 PR 리뷰 요청드립니다.

이번 PR에 직전 PR에 피드백 주신 부분과 Step 1에서 기능 요구사항에 대한 정리가 빠져있어서 해당 작업과 피드백 반영 선행 후에 Step 2 구현을 진행하여 변경사항에 일부 Step 1 내용이 포함되어 있어 참고하시어 리뷰해주시면 감사하겠습니다 !

이번 Step2를 진행하면서 고민했던 부분이 하나 있었는데 리뷰어님 의견이 궁금하여 공유드려봅니다.
- 요구 사항 중 `숫자 이외의 값 또는 음수를 전달하는 경우 RuntimeException 예외를 throw한다.`가 있었는데요, 역할에 따라 클래스를 분리하다보니, Exception을 떨구는 위치가 `StringAddCalculator` 클래스에서 String[]을 탐색하는 것이 아닌, `StringUtil` 클래스의 `parseAsInteger()` 메소드에서 전달받은 String 하나마다 체크하여 Exception을 발생시키는 방법으로 진행했는데, 해당 Exception을 떨구는 역할은 누가 가져가는 게 좋은 구조일지 고민했었습니다.

혹시 정답이 없는 거라면, 리뷰어님은 어느 기준으로 결정하시는 지도 궁금합니다.

그럼 이번 리뷰도 잘 부탁드리겠습니다. :)

stevejkang committed:
fix(step1): apply previous mission feedback

stevejkang committed:
docs(step1): add missing documentation

stevejkang committed:
docs(step2): list up required features and add todo list

stevejkang committed:
feat(calculator): implement string splitter with default/custom rule

- Use `Pattern.compile()` as final
- Add test

stevejkang committed:
feat(calculator): implement string parser

- Add test for `StringParser`

stevejkang committed:
feat(calculator): implement string add calculator

- Add test cases

dhmin5693 commented on README.md:
```diff
@@ -6,4 +6,38 @@
\ No newline at end of file
+* [텍스트와 이미지로 살펴보는 온라인 코드 리뷰 과정](https://github.com/next-step/nextstep-docs/tree/master/codereview)
+
+---
```
정리를 잘 해두셨네요! 😄

리뷰에 앞서서 강의자료에 기재되어 있는 요구조건을 확인합니다.
3단계부터 README 파일에 작성하는 것이 요건이라 1단계에서 별도로 말씀드리진 않았어요.

커밋 내역을 보니 분석하여 잘 작성해두신 README를 바탕으로 코드를 작성한 뒤 체크하여 마무리하신 것 같네요. 💯

stevejkang replied:
프리코스때도 그랬고, 이번에 step 2 하면서도 미리 리드미에 적어보면서 미리 구조를 설계할 수 있어서 실제 구현하는데 시간이 많이 줄어드는 것 같습니다. 앞으로도 꾸준히 한번 진행해보겠습니다 :)

dhmin5693 commented on src/main/java/calculator/utils/StringUtil.java:
```diff
@@ -0,0 +1,7 @@
+
+public class StringUtil {
+    public static boolean isNullOrEmpty(String string) {
+        return string == null || string.length() == 0;
```
public utils 메소드는 테스트를 반드시 작성해서 코드를 보호해주세요.
큰 기능이 아니며 기본적인 부분처럼 보일지라도 만약 이 코드를 변경하게 될 일이 생긴다면 영향을 받는 부분의 안전을 보장할 수 없기 때문이에요.

stevejkang replied:
앗 해당 부분이 빠져있네요 추가하도록 하겠습니다 ! 🙌

dhmin5693 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.stream.Collectors;
+
+public class StringAddCalculator {
+    private static final Integer DEFAULT_RESULT_VALUE = 0;
```
static method만 사용할 목적의 클래스는 외부에서 새 인스턴스를 생성하지 못하도록 제한하는 것을 추천드립니다.

stevejkang replied:
앗 놓쳤던 부분이네요, 피드백 주셔서 감사합니다. 해당 부분은 AssertionError throw 하도록 수정해두었습니다 !

dhmin5693 commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,35 @@
+    private static Integer sumIntegerList(List<Integer> integers) {
+        return integers.stream().reduce(0, Integer::sum);
+    }
+}
```
POSIX 표준에 따라 파일의 끝은 개행 문자로 마무리되어야 합니다.
인텔리제이를 사용하신다면 아래 링크를 참조하여 설정해두시는걸 추천드립니다.
https://stackoverflow.com/questions/16761227/how-to-make-intellij-idea-insert-a-new-line-at-every-end-of-file

stevejkang replied:
항상 습관적으로 마지막 개행문자를 추가하고 있었는데, 인텔리제이 설정으로 있었네요...! 감사합니다 !

깃헙에서는 diff에 자동으로 마지막 개행된 행을 보여주지 않고 wrap해주는 것으로 알고 있습니다. 만약 개행하지 않았을 경우 diff에 아래와 같이 표시되는데, 

![6899e3c0-6ee6-11e9-b1c0-1f6735ff100b_image](https://user-images.githubusercontent.com/47936709/167969796-619b7b63-10f6-4108-ad4a-4500d707fdd7.png)

<img width="229" alt="Screen Shot 2022-05-12 at 9 45 01" src="https://user-images.githubusercontent.com/47936709/167969923-1bee6dbc-3636-4e1e-9aeb-2c6189f5a3c0.png">

지금은 그렇지 않은 것으로 보이는데 한번 더 확인해주실 수 있으실까요? 아니면 제가 놓치는 부분이거나 잘못 알고 있을 수도 있습니다!

dhmin5693 replied:
말씀하신 내용이 맞는 것 같네요.
좋은 내용 하나 배워갑니다. 👍

dhmin5693 commented on src/main/java/calculator/utils/StringSplitter.java:
```diff
@@ -0,0 +1,32 @@
+    private static String[] splitWithRules(String string) {
+        Matcher matcher = CUSTOM_SPLIT_DELIMITER_PATTERN.matcher(string);
+        if (matcher.find()) {
+            return splitByDelimiter(matcher.group(MATCHER_GROUP_VALUE_INDEX), matcher.group(MATCHER_GROUP_CUSTOM_SPLIT_DELIMITER_INDEX));
```
로직을 담는 코드가 한 줄에 담기에는 너무 길어보여요.
적절한 개행은 가독성에 도움이 됩니다.

인텔리제이에서는 한 줄당 100자를 확인할 수 있는 기준선이 있으니 활용해보셔도 좋겠네요.
(**Preferences -> Editor -> Code Style -> Hard Wrap at**)

stevejkang replied:
지금 보니 기준선이 설정되어있음에도 그 값을 넘어간 코드를 작성했네요, 개행 추가해두었습니다 !

dhmin5693 commented on src/test/java/calculator/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,106 @@
+    @DisplayName("Calculator는 null을 받을 경우 0을 반환한다")
+    void calculator_on_null() {
+        Integer sum = StringAddCalculator.splitAndSum(null);
+        assertThat(sum).isEqualTo(0);
```
```suggestion
        assertThat(sum).isZero();
```
0과 비교할 때 사용하는 더 적절한 대안이 있으니 참고바랍니다.

stevejkang replied:
훨씬 읽기 쉬운 방향인 것 같습니다. 제안해주신 방법으로 수정해두겠습니다 !

dhmin5693 commented on src/test/java/calculator/utils/StringParserTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Test
+    @DisplayName("숫자가 아닌 문자열은 오류가 반환되어야 한다")
+    void parse_but_error_on_non_number_value() {
```
요구사항이 아니지만 필요한 케이스라고 생각했는데 포함해주셨네요 💯

dhmin5693 reviewed (changes requested):
안녕하세요.

미션을 너무 잘 수행해주셔서 로직 상에는 딱히 피드백 드릴만한 포인트가 보이지 않네요.
간단한 피드백만 적용해주시면 바로 merge하도록 하겠습니다.

예외를 어느 위치에서 발생시키는게 좋은 구조냐는 질문을 주셨어요.
당연한 말이지만 코드를 작성하는 방법에는 정답이 없습니다.
코드의 좋고 나쁨은 정량적으로 평가할 수 없습니다.
현재 상황에서 활용할 수 있는 가장 좋은 방법을 적용한 코드가 가장 좋아요.

클린코드를 지향해서 너무 잘게잘게 쪼개놨더니 팀원 사이에서 코드 읽기가 어렵다는 반응이 나온다면 어떨까요?
좋은 코드일까요? 적당한 선에서의 합의를 본 코드가 더 나을 수 있는거죠.
이런 상황에서 클린코드가 좋은 코드가 되려면 팀원들과 합의 후 점진적으로 개선해나가는 과정이 필요합니다.

제가 이 미션을 진행한다면 예외의 위치는 준영님과 크게 다르지 않을 것 같아요.
현재 코드의 논리 구조를 평가해보면 준영님께서 예외를 발생시킨 위치는 굉장히 적절하거든요.

- 문자열을 받아서 숫자의 합계를 도출한다.
    1. 문자열을 구분자 기준으로 분할한다.
    2. 분할한 문자열을 숫자형으로 변환한다.
    3. 변환된 숫자를 더하여 반환한다.

위의 논리 구조로 코드를 작성하셨는데요, 변환의 책임은 2번에서 가져가고 있습니다.
당연히 변환이 안 된 경우에 대한 예외도 2번에서 발생하는게 합리적이겠죠.
제 기준은 해당 책임을 갖는 곳에서 예외도 처리한다고 말씀드릴 수 있겠습니다.

stevejkang committed:
fix(calculator): apply step2 feedbacks

- Add missing test for `StringUtil`
- To improve readability, replace `isEqualTo(0)` to `isZero()`
- Add line break when code is too long
- Throw `AssertionError` to prevent to be an instance when the class has only static methods

dhmin5693 reviewed (approved):
피드백 적용하시느라 수고 많으셨습니다.
다음 미션에서 뵈어요 :)
